minigui/layer.ts

```diff
@@ -0,0 +1,407 @@
+    return false;
+  }
+}
+
+abstract class DataLayer extends Layer {
+  constructor(board: Board, private dataPropName: string) {
+    super(board);
+  }
+
+  protected getData<T>(dataObj: any): DataProp<T> {
+    let prop: any = dataObj[this.dataPropName];
+    if (prop === undefined) {
+      return undefined;
+    }
+    return prop as T | null;
+  }
+}
+
+class Grid extends StaticLayer {
+  private style = '#864';
+
+  draw() {
+    let starPointRadius = Math.min(4, Math.max(this.board.stoneRadius / 10, 2.5));
+    let ctx = this.board.ctx;
+    let size = this.board.size;
+    let pr = pixelRatio();
+
+    ctx.strokeStyle = this.style;
+    ctx.lineWidth = pr;
+    ctx.lineCap = 'round';
+
+    ctx.beginPath();
+    for (let i = 0; i < size; ++i) {
+      let left = this.boardToCanvas(i, 0);
+      let right = this.boardToCanvas(i, size - 1);
+      let top = this.boardToCanvas(0, i);
+      let bottom = this.boardToCanvas(size - 1, i);
+      ctx.moveTo(0.5 + Math.round(left.x), 0.5 + Math.round(left.y));
+      ctx.lineTo(0.5 + Math.round(right.x), 0.5 + Math.round(right.y));
+      ctx.moveTo(0.5 + Math.round(top.x), 0.5 + Math.round(top.y));
+      ctx.lineTo(0.5 + Math.round(bottom.x), 0.5 + Math.round(bottom.y));
+    }
+    ctx.stroke();
+
+    // Draw star points.
+    ctx.fillStyle = this.style;
+    ctx.strokeStyle = '';
+    for (let p of STAR_POINTS[size]) {
+      let c = this.boardToCanvas(p[0], p[1]);
+      ctx.beginPath();
+      ctx.arc(c.x + 0.5, c.y + 0.5, starPointRadius * pr, 0, 2 * Math.PI);
+      ctx.fill();
+    }
+  }
+}
+
+class Label extends StaticLayer {
+  draw() {
+    let ctx = this.board.ctx;
+    let size = this.board.size;
+
+    let textHeight = Math.floor(0.3 * this.board.stoneRadius);
+    ctx.font = `${textHeight}px sans-serif`;
+    ctx.fillStyle = '#9d7c4d';
+
+    // Draw column labels.
+    ctx.textAlign = 'center';
+    ctx.textBaseline = 'alphabetic';
+    for (let i = 0; i < size; ++i) {
+      let c = this.boardToCanvas(-0.66, i);
+      ctx.fillText(COL_LABELS[i], c.x, c.y);
+    }
+    ctx.textBaseline = 'top';
+    for (let i = 0; i < size; ++i) {
+      let c = this.boardToCanvas(size - 0.33, i);
+      ctx.fillText(COL_LABELS[i], c.x, c.y);
+    }
+
+    // Draw row labels.
+    ctx.textAlign = 'right';
+    ctx.textBaseline = 'middle';
+    for (let i = 0; i < size; ++i) {
+      let c = this.boardToCanvas(i, -0.66);
+      ctx.fillText((size - i).toString(), c.x, c.y);
+    }
+    ctx.textAlign = 'left';
+    for (let i = 0; i < size; ++i) {
+      let c = this.boardToCanvas(i, size - 0.33);
+      ctx.fillText((size - i).toString(), c.x, c.y);
+    }
+  }
+}
+
+class Caption extends StaticLayer {
+  constructor(board: Board, private caption: string) {
+    super(board);
+  }
+
+  draw() {
+    let ctx = this.board.ctx;
+    let size = this.board.size;
+
+    let textHeight = Math.floor(0.4 * this.board.stoneRadius);
+    ctx.font = `${textHeight}px sans-serif`;
```

Comment:
this syntax is kind of amazing.

Reply:
Yeah, I love it. It's one of my favourite things about modern JavaScript